File: laravel/tests/SignTest.php
<?php

declare(strict_types=1);

namespace X402\Laravel\Tests;

use PHPUnit\Framework\TestCase;
use X402\Laravel\Sign;

final class SignTest extends TestCase
{
    /** @var array<string,mixed> */
    private static array $vectors;

    public static function setUpBeforeClass(): void
    {
        self::$vectors = json_decode(
            (string) file_get_contents(
                dirname(__DIR__, 3) . '/docs/api/x402-conformance-vectors.json'
            ),
            true,
            512,
            JSON_THROW_ON_ERROR
        );
    }

    public function testSchemeIsFrozen(): void
    {
        $this->assertSame('X402v1', Sign::SCHEME);
    }

    public function testCanonicalIsTheFrozenLayout(): void
    {
        $c = Sign::canonical('post', '/api/v1/verify', 1700000000, 'nonce-1', '{"a":1}');
        $expected = "X402v1\nPOST\n/api/v1/verify\n1700000000\nnonce-1\n" . hash('sha256', '{"a":1}');
        $this->assertSame($expected, $c);
        $this->assertSame(self::$vectors['kat']['canonical'], $c);
    }

    public function testEmptyBodyHash(): void
    {
        $c = Sign::canonical('GET', '/api/v1/challenge', 1, 'n', '');
        $this->assertStringEndsWith("\n" . self::$vectors['body_hashes'][0]['sha256'], $c);
    }

    public function testKatSignatureMatchesFrozenVector(): void
    {
        $k = self::$vectors['kat'];
        $got = Sign::sign($k['secret'], $k['method'], $k['path'], $k['timestamp'], $k['nonce'], $k['body']);
        $this->assertMatchesRegularExpression('/^[0-9a-f]{64}$/', $got);
        $this->assertSame($k['signature'], $got);
    }

    public function testUnicodeBodyHashesRawUtf8Bytes(): void
    {
        $body = '{"name":"こんにちは"}';
        $c = Sign::canonical('POST', '/api/v1/verify', 1700000000, 'n', $body);
        $this->assertStringEndsWith("\n" . hash('sha256', $body), $c);
    }

    public function testNonIntTimestampRejected(): void
    {
        $this->expectException(\TypeError::class);
        // @phpstan-ignore-next-line — intentionally wrong type under strict_types
        Sign::canonical('POST', '/p', '1700000000', 'n', '');
    }

    public function testSignedHeadersShape(): void
    {
        $h = Sign::signedHeaders('x402_test_abc', 'x402sk_test_x', 'POST', '/api/v1/challenge', '{"route":"/p"}');
        $this->assertSame('x402_test_abc', $h['X-X402-Key']);
        $this->assertMatchesRegularExpression('/^\d+$/', $h['X-X402-Timestamp']);
        $this->assertGreaterThanOrEqual(16, strlen($h['X-X402-Nonce']));
        $this->assertMatchesRegularExpression('/^[0-9a-f]{64}$/', $h['X-X402-Signature']);
        $ts = (int) $h['X-X402-Timestamp'];
        $this->assertSame(
            Sign::sign('x402sk_test_x', 'POST', '/api/v1/challenge', $ts, $h['X-X402-Nonce'], '{"route":"/p"}'),
            $h['X-X402-Signature']
        );
    }

    public function testSignedHeadersNoncesUnique(): void
    {
        $a = Sign::signedHeaders('k', 's', 'POST', '/api/v1/challenge', '');
        $b = Sign::signedHeaders('k', 's', 'POST', '/api/v1/challenge', '');
        $this->assertNotSame($a['X-X402-Nonce'], $b['X-X402-Nonce']);
    }

    public function testSharedVectorsFullConformance(): void
    {
        foreach (self::$vectors['body_hashes'] as $v) {
            $this->assertSame($v['sha256'], hash('sha256', $v['body']));
        }
        $k = self::$vectors['kat'];
        $this->assertSame($k['canonical'], Sign::canonical($k['method'], $k['path'], $k['timestamp'], $k['nonce'], $k['body']));
        $this->assertSame($k['signature'], Sign::sign($k['secret'], $k['method'], $k['path'], $k['timestamp'], $k['nonce'], $k['body']));
    }
}

File: go/sign_test.go
package x402

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"testing"
)

type vectors struct {
	KAT struct {
		Secret    string `json:"secret"`
		Method    string `json:"method"`
		Path      string `json:"path"`
		Timestamp int64  `json:"timestamp"`
		Nonce     string `json:"nonce"`
		Body      string `json:"body"`
		BodySHA   string `json:"body_sha256"`
		Canonical string `json:"canonical"`
		Signature string `json:"signature"`
	} `json:"kat"`
	BodyHashes []struct {
		Body   string `json:"body"`
		SHA256 string `json:"sha256"`
	} `json:"body_hashes"`
}

func loadVectors(t *testing.T) vectors {
	t.Helper()
	p := filepath.Join("..", "..", "docs", "api", "x402-conformance-vectors.json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("read shared vectors: %v", err)
	}
	var v vectors
	if err := json.Unmarshal(b, &v); err != nil {
		t.Fatalf("parse shared vectors: %v", err)
	}
	return v
}

func TestSchemeIsFrozen(t *testing.T) {
	if Scheme != "X402v1" {
		t.Fatalf("Scheme = %q, want X402v1", Scheme)
	}
}

func TestCanonicalIsFrozenLayout(t *testing.T) {
	v := loadVectors(t)
	got := Canonical("post", "/api/v1/verify", 1700000000, "nonce-1", `{"a":1}`)
	sum := sha256.Sum256([]byte(`{"a":1}`))
	want := "X402v1\nPOST\n/api/v1/verify\n1700000000\nnonce-1\n" + hex.EncodeToString(sum[:])
	if got != want {
		t.Fatalf("canonical mismatch\n got=%q\nwant=%q", got, want)
	}
	if got != v.KAT.Canonical {
		t.Fatalf("canonical != shared vector\n got=%q\nwant=%q", got, v.KAT.Canonical)
	}
}

func TestEmptyBodyHash(t *testing.T) {
	v := loadVectors(t)
	got := Canonical("GET", "/api/v1/challenge", 1, "n", "")
	if !regexpEndsWith(got, "\n"+v.BodyHashes[0].SHA256) {
		t.Fatalf("empty-body hash suffix mismatch: %q", got)
	}
}

func TestKATSignatureMatchesSharedVector(t *testing.T) {
	v := loadVectors(t)
	got := Sign(v.KAT.Secret, v.KAT.Method, v.KAT.Path, v.KAT.Timestamp, v.KAT.Nonce, v.KAT.Body)
	if !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(got) {
		t.Fatalf("signature not lowercase-hex-64: %q", got)
	}
	if got != v.KAT.Signature {
		t.Fatalf("KAT signature mismatch\n got=%s\nwant=%s", got, v.KAT.Signature)
	}
}

func TestUnicodeBodyHashesRawUTF8(t *testing.T) {
	body := `{"name":"こんにちは"}`
	got := Canonical("POST", "/api/v1/verify", 1700000000, "n", body)
	sum := sha256.Sum256([]byte(body))
	if !regexpEndsWith(got, "\n"+hex.EncodeToString(sum[:])) {
		t.Fatalf("unicode body hash mismatch: %q", got)
	}
}

func TestSignedHeadersShape(t *testing.T) {
	h := SignedHeaders("x402_test_abc", "x402sk_test_x", "POST", "/api/v1/challenge", `{"route":"/p"}`)
	if h["X-X402-Key"] != "x402_test_abc" {
		t.Fatalf("key header = %q", h["X-X402-Key"])
	}
	if !regexp.MustCompile(`^\d+$`).MatchString(h["X-X402-Timestamp"]) {
		t.Fatalf("timestamp not integer: %q", h["X-X402-Timestamp"])
	}
	if len(h["X-X402-Nonce"]) < 16 {
		t.Fatalf("nonce too short: %q", h["X-X402-Nonce"])
	}
	if !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(h["X-X402-Signature"]) {
		t.Fatalf("signature not hex64: %q", h["X-X402-Signature"])
	}
}

func TestSignedHeadersNoncesUnique(t *testing.T) {
	a := SignedHeaders("k", "s", "POST", "/api/v1/challenge", "")
	b := SignedHeaders("k", "s", "POST", "/api/v1/challenge", "")
	if a["X-X402-Nonce"] == b["X-X402-Nonce"] {
		t.Fatal("nonces not unique across calls")
	}
}

func TestSharedVectorsFullConformance(t *testing.T) {
	v := loadVectors(t)
	for _, bh := range v.BodyHashes {
		sum := sha256.Sum256([]byte(bh.Body))
		if hex.EncodeToString(sum[:]) != bh.SHA256 {
			t.Fatalf("body hash mismatch for %q", bh.Body)
		}
	}
	if Canonical(v.KAT.Method, v.KAT.Path, v.KAT.Timestamp, v.KAT.Nonce, v.KAT.Body) != v.KAT.Canonical {
		t.Fatal("canonical drift vs shared vector")
	}
	if Sign(v.KAT.Secret, v.KAT.Method, v.KAT.Path, v.KAT.Timestamp, v.KAT.Nonce, v.KAT.Body) != v.KAT.Signature {
		t.Fatal("signature drift vs shared vector")
	}
}

func regexpEndsWith(s, suffix string) bool {
	return len(s) >= len(suffix) && s[len(s)-len(suffix):] == suffix
}
